[VectorCombine] Combine scalar fneg with insert/extract to vector fneg when length is different

ParkHanbum opened the pull request:
insertelt DestVec, (fneg (extractelt SrcVec, Index)), Index
-> shuffle DestVec, (shuffle (fneg SrcVec), poison, SrcMask), Mask

Original combining left the combine between vectors of different
lengths as a TODO. this commit do that. (see #[baab4aa])

ParkHanbum committed:
Add tests for combine extract/insert between vectors of different lengths

ParkHanbum committed:
[VectorCombine] Combine scalar fneg with insert/extract to vector fneg when length is different

insertelt DestVec, (fneg (extractelt SrcVec, Index)), Index
-> shuffle DestVec, (shuffle (fneg SrcVec), poison, SrcMask), Mask

Original combining left the combine between vectors of different
lengths as a TODO. this commit do that. (see #[baab4aa])

ParkHanbum commented:
can I request review?

mshockwave commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -698,14 +698,31 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
+    Value *LenChgShuf = Builder.CreateShuffleVector(
+        SrcVec, PoisonValue::get(SrcVecTy), SrcMask);
+    NewShuf = Builder.CreateShuffleVector(DestVec, LenChgShuf, Mask);
+  } else
```
please match the curly braces. https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

ParkHanbum committed:
fix mis-matched braces

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -682,7 +682,8 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
   Mask[Index] = Index + NumElts;
 
-  Type *ScalarTy = VecTy->getScalarType();
+  Type *ScalarTy = SrcVecTy->getScalarType();
```
Move this up and avoid the repeated SrcVecTy->getScalarType()

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -682,7 +682,8 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
 
-  Type *ScalarTy = VecTy->getScalarType();
+  Type *ScalarTy = SrcVecTy->getScalarType();
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
```
Remove this (CostKind is now a VectorCombine class member)

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -697,14 +698,32 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
+  // we need to add a length-change vector. Add this cost.
+  if (NeedLenChg)
+    NewCost +=
+        TTI.getShuffleCost(TargetTransformInfo::SK_Select, SrcVecTy, Mask);
```
Shouldn't be a Select shuffle kind? Create the SrcMask below earlier and use SingleSrc SK.

RKSimon reviewed (commented):
Sorry I missed this.

ParkHanbum committed:
remove costkind

ParkHanbum committed:
move ScalarTy to reduce duplicate

ParkHanbum committed:
more accurate shufflecost with PermuteSingleSrc

ParkHanbum committed:
updated testcase

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -697,14 +696,34 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
+  if (NeedLenChg) {
+    SrcMask.assign(NumElts, PoisonMaskElem);
+    NewCost += TTI.getShuffleCost(TargetTransformInfo::SK_PermuteSingleSrc,
+                                  SrcVecTy, SrcMask, CostKind);
```
SrcMask is all PoisonMaskElem - move the "SrcMask[Index] = Index;" line?

ParkHanbum replied:
I'm asking because I don't know, does it affect the cost calculation?

ParkHanbum committed:
move indexing of SrcMask

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -665,9 +665,10 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
-  if (SrcVec->getType() != VecTy)
+  auto *SrcVecTy = cast<FixedVectorType>(SrcVec->getType());
+  auto *ScalarTy = SrcVecTy->getScalarType();
+  if (ScalarTy != VecTy->getScalarType())
```
We can't guarantee that SrcVec is a FixedVectorType, you're probably safer doing this:
```cpp
auto *ScalarTy = VecTy->getScalarType();
auto *SrcVecTy = dyn_cast<FixedVectorType>(SrcVec->getType());
if (!SrcVecTy || ScalarTy != SrcVecTy->getScalarType())
```

RKSimon commented on llvm/test/Transforms/VectorCombine/X86/extract-fneg-insert.ll:
```diff
@@ -46,6 +72,25 @@ define <2 x double> @ext1_v2f64(<2 x double> %x, <2 x double> %y) {
+  %n = fneg nsz double %e
+  %r = insertelement <4 x double> %y, double %n, i32 1
+  ret <4 x double> %r
+}
```
Please can you add coverage from when you extract from an index greater than the vector width of the destination:
```
  %e = extractelement <4 x double> %x, i32 3
  %n = fneg nsz double %e
  %r = insertelement <2 x double> %y, double %n, i32 1
  ret <4 x double> %r
```

ParkHanbum replied:
done. but this test case didn't work as properly because we restrict of extract/insert index to be same.

ParkHanbum committed:
add checking SrcVec is FixedVectorType

ParkHanbum committed:
add test for extract from an index greater than the vector width of the destination

ParkHanbum committed:
Fix misused of SrcVecTy instead VecTy

ParkHanbum committed:
Call CreateShuffleVector without the poison

If we call CreateShuffleVector with only one value as
an argument, it will create a poison vector internally
and it shuffled.

ParkHanbum commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -665,9 +665,10 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
   auto *VecTy = cast<FixedVectorType>(I.getType());
-  if (SrcVec->getType() != VecTy)
+  auto *SrcVecTy = dyn_cast<FixedVectorType>(SrcVec->getType());
+  auto *ScalarTy = SrcVecTy->getScalarType();
```
it was terrible mistake. so sorry.

RKSimon reviewed (approved):
LGTM - cheers

RKSimon commented:
@ParkHanbum Do you have commit access?

ParkHanbum commented:
@RKSimon no I haven't

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -697,14 +696,33 @@ bool VectorCombine::foldInsExtFNeg(Instruction &I) {
-  replaceValue(I, *Shuf);
+  if (NeedLenChg) {
+    // shuffle DestVec, (shuffle (fneg SrcVec), poison, SrcMask), Mask
+    Value *LenChgShuf = Builder.CreateShuffleVector(SrcVec, SrcMask);
```
@ParkHanbum - this is regression typo - sorry I missed this :( 
```
Value *LenChgShuf = Builder.CreateShuffleVector(VecFNeg, SrcMask);
```

ParkHanbum replied:
I'm sorry for my mistake.. Can I send a PR again?
